crates/trigger-http/src/handler.rs

```diff
@@ -320,28 +316,36 @@ impl HttpHandlerExecutor {
 }
 
 /// Whether this handler uses the custom Spin http handler interface for wasi-http
-enum HandlerType {
+#[derive(Copy, Clone)]
+pub enum HandlerType {
     Spin,
-    Wasi,
+    Wasi0_2,
+    Wasi2023_11_10,
+    Wasi2023_10_18,
 }
 
 const WASI_HTTP_EXPORT_2023_10_18: &str = "wasi:http/incoming-handler@0.2.0-rc-2023-10-18";
 const WASI_HTTP_EXPORT_2023_11_10: &str = "wasi:http/incoming-handler@0.2.0-rc-2023-11-10";
 const WASI_HTTP_EXPORT_0_2_0: &str = "wasi:http/incoming-handler@0.2.0";
 
 impl HandlerType {
-    /// Determine the handler type from the exports
-    fn from_exports(mut exports: wasmtime::component::Exports<'_>) -> Option<HandlerType> {
-        if exports.instance(WASI_HTTP_EXPORT_2023_10_18).is_some()
-            || exports.instance(WASI_HTTP_EXPORT_2023_11_10).is_some()
-            || exports.instance(WASI_HTTP_EXPORT_0_2_0).is_some()
-        {
-            return Some(HandlerType::Wasi);
-        }
-        if exports.instance("fermyon:spin/inbound-http").is_some() {
-            return Some(HandlerType::Spin);
+    /// Determine the handler type from the exports of a component
+    pub fn from_component(ty: &ComponentType) -> Result<HandlerType> {
+        for (name, _) in ty.exports() {
+            match name {
+                WASI_HTTP_EXPORT_2023_10_18 => return Ok(HandlerType::Wasi2023_10_18),
+                WASI_HTTP_EXPORT_2023_11_10 => return Ok(HandlerType::Wasi2023_11_10),
+                WASI_HTTP_EXPORT_0_2_0 => return Ok(HandlerType::Wasi0_2),
+                "fermyon:spin/inbound-http" => return Ok(HandlerType::Spin),
+                _ => {}
```

Comment:
I think this isn't equivalent to the old code if a component exports more than one of these interfaces. I'm not sure whether that is an issue in practice. @dicej @rylev thoughts?

Reply:
Looks like the old code uses the same order, so I think it would behave the same in that (unusual) case.

Reply:
If I'm reading correctly, for an app that exports both spin and wasi interfaces the old code will always return wasi while the new code will depend on the order of `exports()` (presumably the order of items in the binary).

Reply:
That's right yeah, I can preserve the exact same behavior as before if desired, but I figured that level of complexity probably wasn't what was intended here

Reply:
As long as tooling isn't "likely" to be emitting multiple of these exported interfaces I'm fine with the code as-is. I'm mostly just unsure about spin-componentize.

Reply:
I think we're unlikely to run into this situation with components built using Spin provided tooling (e.g., the SDKs). However, it does seem like we'd not want to rely on order of exports to determine which will be invoked, and indeed there should be a preferred ordering of wasi 0.2, wasi release candidates, Spin specific.

Reply:
I've pushed up a commit to generate an error if more than one handler is found to help resolve this